docs/codeql/codeql-language-guides/analyzing-data-flow-in-swift.rst

```diff
@@ -0,0 +1,289 @@
+Like local data flow, a predicate ``localTaintStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo)`` holds if there is an immediate taint propagation edge from the node ``nodeFrom`` to the node ``nodeTo``.
+You can apply the predicate recursively, by using the ``+`` and ``*`` operators, or you can use the predefined recursive predicate ``localTaint``.
+
+For example, you can find taint propagation from an expression ``source`` to an expression ``sink`` in zero or more local steps:
+
+.. code-block:: ql
+
+     TaintTracking::localTaint(DataFlow::exprNode(source), DataFlow::exprNode(sink))
+
+Examples of local data flow
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This query finds the ``format`` argument passed into each call to ``String.init(format:_:)``:
+
+.. code-block:: ql
+
+    import swift
+
+    from CallExpr call, MethodDecl method
+    where
+      call.getStaticTarget() = method and
+      method.hasQualifiedName("String", "init(format:_:)")
+    select call.getArgument(0).getExpr()
+
+Unfortunately this will only give the expression in the argument, not the values which could be passed to it.
+So we use local data flow to find all expressions that flow into the argument:
+
+.. code-block:: ql
+
+    import swift
+    import codeql.swift.dataflow.DataFlow
+
+    from CallExpr call, MethodDecl method, Expr sourceExpr, Expr sinkExpr
+    where
+      call.getStaticTarget() = method and
+      method.hasQualifiedName("String", "init(format:_:)") and
+      sinkExpr = call.getArgument(0).getExpr() and
+      DataFlow::localFlow(DataFlow::exprNode(sourceExpr), DataFlow::exprNode(sinkExpr))
+    select sourceExpr, sinkExpr
+
+We can vary the source, for example, making the source the parameter of a function rather than an expression. The following query finds where a parameter is used for the format:
+
+.. code-block:: ql
+
+    import swift
+    import codeql.swift.dataflow.DataFlow
+
+    from CallExpr call, MethodDecl method, ParamDecl sourceParam, Expr sinkExpr
+    where
+      call.getStaticTarget() = method and
+      method.hasQualifiedName("String", "init(format:_:)") and
+      sinkExpr = call.getArgument(0).getExpr() and
+      DataFlow::localFlow(DataFlow::parameterNode(sourceParam), DataFlow::exprNode(sinkExpr))
+    select sourceParam, sinkExpr
+
+The following example finds calls to ``String.init(format:_:)`` where the format string is not a hard-coded string literal:
+
+.. code-block:: ql
+
+    import swift
+    import codeql.swift.dataflow.DataFlow
+
+    from CallExpr call, MethodDecl method, Expr sinkExpr
+    where
+      call.getStaticTarget() = method and
+      method.hasQualifiedName("String", "init(format:_:)") and
+      sinkExpr = call.getArgument(0).getExpr() and
+      not exists(StringLiteralExpr sourceLiteral |
+        DataFlow::localFlow(DataFlow::exprNode(sourceLiteral), DataFlow::exprNode(sinkExpr))
+      )
+    select call, "Format argument to " + method.getName() + " isn't hard-coded."
```

Comment:
This will have a FP when there's no dataflow node for `sinkExpr` (for example, if the expression isn't reachable because the control-flow library determined that the code was dead). We have observed this in C++, which is why we recently rewrote this example query when we updated the documentation with use-use flow (see https://codeql.github.com/docs/codeql-language-guides/analyzing-data-flow-in-cpp-new/#examples). Based on the way we wrote it for C++, I suggest something like this:

```suggestion
    from CallExpr call, MethodDecl method, DataFlow::Node sinkNode
    where
      call.getStaticTarget() = method and
      method.hasQualifiedName("String", "init(format:_:)") and
      sinkNode.asExpr() = call.getArgument(0).getExpr() and
      not exists(StringLiteralExpr sourceLiteral |
        DataFlow::localFlow(DataFlow::exprNode(sourceLiteral), sinkNode)
      )
    select call, "Format argument to " + method.getName() + " isn't hard-coded."
```

Feel free to modify this to use `DataFlow::exprNode` instead of `asExpr` if you think that is more pedagogically helpful.

Reply:
Thanks for this subtle correction, and yes I'm happy with `DataFlow::Node` and `asExpr` they're already discussed in the doc so should be familiar to the reader.

Reply:
I've created a merge conflict with your suggested change, so I've re-implemented it manually.  (done)